### PR TITLE
Mark 4, 5B stream reader's read method now can set invalid data `fill_value`

### PR DESCRIPTION
**Pull request by cczhu**
Small bugfix (to #101) to allow the user to set the `fill_value` for invalid data in Mark 4 and 5B stream readers.  Implemented associated tests (which write multiple frames because the user currently cannot set `fill_value` upon reader initialization).

### CODE REVIEW

File: baseband/mark4/base.py [hunk 1]
```diff
@@ -389,14 +389,15 @@ def read(self, count=None, fill_value=0., out=None):
...
-    def _read_frame(self):
+    def _read_frame(self, fill_value=0.):
         frame_nr = self.offset // self.samples_per_frame
         self.fh_raw.seek(self.offset0 + frame_nr * self.header0.framesize)
         self._frame = self.read_frame(ntrack=self.header0.ntrack,
                                       decade=self.header0.decade)
         # Convert payloads to data array.
         self._frame_data = self._frame.data
         self._frame_nr = frame_nr
+        self._frame.invalid_data_value = fill_value
```
Comment (mhvk) [inline baseband/mark4/base.py]: I think you need to move this up above the `.data` call, since that does the actual decoding and invalid data setting.

File: baseband/mark4/frame.py [hunk 2]
```diff
@@ -69,6 +69,7 @@ class Mark4Frame(VLBIFrameBase):
 
     _header_class = Mark4Header
     _payload_class = Mark4Payload
+    invalid_data_value = 0.
```
Comment (mhvk) [inline baseband/mark4/frame.py]: Why is this needed? Doesn't it inherit it from `VLBIFrameBase`?
... though possibly this is a good way for it to appear in the docs...

File: baseband/mark4/base.py [hunk 3]
```diff
@@ -374,7 +374,7 @@ def read(self, count=None, fill_value=0., out=None):
             frame_nr, sample_offset = divmod(self.offset,
                                              self.samples_per_frame)
             if frame_nr != self._frame_nr:
-                self._read_frame()
+                self._read_frame(fill_value)
 
             data = self._frame.data
```
Comment (mhvk) [inline baseband/mark4/base.py]: Actually, I was wrong below, since here it just gets `_frame.data` again instead of using `_frame_data`. Since neither payload nor frame cache the decoded data, I think we should use `_frame_data` here.

File: baseband/mark5b/base.py [hunk 4]
```diff
@@ -258,14 +258,15 @@ def read(self, count=None, fill_value=0., out=None):
...
-    def _read_frame(self):
+    def _read_frame(self, fill_value=0.):
         self.fh_raw.seek(self.offset // self.samples_per_frame *
                          self._frame.size)
         self._frame = self.read_frame(ref_mjd=self.header0.kday,
                                       nchan=self._sample_shape.nchan,
                                       bps=self.bps)
         # Convert payloads to data array.
         self._frame_data = self._frame.data
+        self._frame.invalid_data_value = fill_value
```
Comment (mhvk) [inline baseband/mark5b/base.py]: Same as for mark4 (both moving this up and using `self._frame_data` in `read()`)

File: baseband/mark4/frame.py (hunk 2 above)
Comment (cczhu) [reply baseband/mark4/frame.py]: Ahh, yeah, I'll remove it. It's only in VDIF's `frame.py` because of `frameset`, but I got confused.

File: baseband/mark4/base.py (hunk 3 above)
Comment (cczhu) [reply baseband/mark4/base.py]: Removing that prevents decoding the first frame by the stream reader.  We could add `self._frame_data = self._frame.data` in `__init__`, but I think that keeping `data = self._frame.data` in `read` while removing it in `_read_frame` is a more elegant (if less efficient) solution and makes the code more symmetric with VDIF.  See further comments below.

**Comment by cczhu** [comment]
I didn't appreciate that data is not automatically decoded upon using `read_frame`!  This has inspired me to make all the VLBI stream readers decode only within `read` (and not in `_read_frame`), setting `self._frame.invalid_data_value = fill_value` before decoding.  This partly resolves #106 for all VLBI classes.  It doesn't resolve switching `fill_value` in the middle of reading a VDIF frame (hopefully no one is doing that!), since those values are still cached by `VDIFFrameSet`.

I do agree that re-decoding a frame's worth of data each time `read` is called is inefficient.  We should think up a better solution than this after we profile Baseband.

File: baseband/mark4/base.py [hunk 5]
```diff
@@ -376,6 +376,9 @@ def read(self, count=None, fill_value=0., out=None):
             if frame_nr != self._frame_nr:
                 self._read_frame()
 
+            # Set decoded value for invalid data.
+            self._frame.invalid_data_value = fill_value
+            # Decode data into array.
             data = self._frame.data
```
Comment (mhvk) [inline baseband/mark4/base.py]: We can make the decoding somewhat less inefficient by using the fact that one can slice the frame, in which case only part will be decoded; so remove this one and rewrite as
```
nsample = ...
selection = slice(sample_offset, sample_offset + nsample)
if self.thread_ids:
    selection = (selection, self.thread_ids)
sample = self.offset - offset0
result[sample:sample + nsample] = self._frame[selection]
```

File: baseband/mark4/frame.py (hunk 2 above)
Comment (mhvk) [reply baseband/mark4/frame.py]: Still to be removed?

File: baseband/mark4/base.py (hunk 5 above)
Comment (cczhu) [reply baseband/mark4/base.py]: Let's leave this for a follow-up PR (probably while looking at performance in general).
